### PR TITLE
Remove BinaryFormatter from GetSDKReferenceFiles

### PR DESCRIPTION
**Pull request by Forgind**
I think this is almost right, but TranslatorHelpers needs System.Collections.Concurrent on .NET 3.5, or I need to be able to translate a concurrent dictionary without translating it. (?) Any idea on how I can get around that?

### CODE REVIEW

**Comment by KirillOsenkov** [comment]
I would hardcode the translation logic for each type T that we need, and just write the count of key-value pairs, followed by each key and value separately.

**Comment by Forgind** [comment]
Can you explain why we need the counts? I haven't looked at how complete our test coverage of this task was, but it seems to pass whatever tests we previously had without it.

**Comment by KirillOsenkov** [comment]
I might be not understanding it well, but how are you going to read the keys and values if you don’t know how many there are?

**Comment by Forgind** [comment]
Looks like the answer is: there weren't tests. Makes it easy to pass them 😋

I was imagining something like looking for closing elements. In json, for instance, they have { "key1": "val1", "key2": "val2" }. You don't need a count because the opening and closing braces match. I don't know if there's an equivalent here, but I now think it just takes references and gets or sets them based on what sort of data you ask for, so since the dictionary would have started out full in serializing, it would serialize properly. Then, in deserialization, the dictionary would be _null_, and it would throw an exception. If I'd initialized the dictionary, it would still be empty and remain so. I added a count. I should probably add tests.

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 1]
```diff
@@ -1205,7 +1194,7 @@ private static IEnumerable<string> GetAllReferenceDirectories(string sdkRoot)
         /// </summary>
         /// <remarks>This is a serialization format. Do not change member naming.</remarks>
         [Serializable]
-        private class SdkReferenceInfo
+        internal class SdkReferenceInfo
         {
             /// <summary>
             /// Constructor
```
Comment (BenVillalobos) [inline src/Tasks/GetSDKReferenceFiles.cs:1200]: While you're getting rid of useless summaries, might as well delete this one.

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 2]
```diff
@@ -1246,63 +1221,105 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
+            internal ConcurrentDictionary<string, SdkReferenceInfo> PathToReferenceMetadata { get { return pathToReferenceMetadata; } }
 
             /// <summary>
             /// Dictionary which maps a directory to a list of file names within that directory. This is used to shortcut hitting the disk for the list of files inside of it.
             /// </summary>
-            public ConcurrentDictionary<string, List<string>> DirectoryToFileList { get; }
+            public ConcurrentDictionary<string, List<string>> DirectoryToFileList { get { return directoryToFileList; } }
 
             /// <summary>
             /// Hashset
```
Comment (BenVillalobos) [inline src/Tasks/GetSDKReferenceFiles.cs:1263]: nit: remove useless summary

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 3]
```diff
@@ -1246,63 +1221,105 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
+                hash = 0;
+            }
+
+            public SDKInfo(ITranslator translator) : this()
+            {
+                Translate(translator);
+            }
 
             /// <summary>
             /// Constructor
```
Comment (BenVillalobos) [inline src/Tasks/GetSDKReferenceFiles.cs:1243]: nit: remove useless summary

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 4]
```diff
@@ -1246,63 +1221,105 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
+                this.pathToReferenceMetadata = pathToReferenceMetadata;
+                this.directoryToFileList = directoryToFileList;
+                this.hash = cacheHash;
             }
 
             /// <summary>
             /// A dictionary which maps a file path to a structure that contain some metadata information about that file.
             /// </summary>
-            public ConcurrentDictionary<string, SdkReferenceInfo> PathToReferenceMetadata { get; }
+            internal ConcurrentDictionary<string, SdkReferenceInfo> PathToReferenceMetadata { get { return pathToReferenceMetadata; } }
```
Comment (BenVillalobos) [inline src/Tasks/GetSDKReferenceFiles.cs:1255]: Why is this changed from public to internal, but other properties untouched?

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 5]
```diff
@@ -1246,63 +1221,105 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
+                    dictionary[key] = value;
+                }                
             }
         }
 
         /// <summary>
         /// This class represents the context information used by the background cache serialization thread.
         /// </summary>
-        private class SaveContext
+        internal class SaveContext
```
Comment (BenVillalobos) [inline src/Tasks/GetSDKReferenceFiles.cs:1322]: Why did this need to change?

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 6]
```diff
@@ -1246,63 +1221,105 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
 
-                    return info;
+        private static void TranslateConcurrentDictionary<T>(ITranslator translator, ref ConcurrentDictionary<string, T> dictionary, ObjectTranslator<T> objTranslator)
+        {
+            int count = dictionary.Count;
+            translator.Translate(ref count);
+            string[] keys = dictionary.Keys.ToArray();
+            for (int i = 0; i < count; i++)
+            {
+                if (i < keys.Length)
```
Comment (BenVillalobos) [inline src/Tasks/GetSDKReferenceFiles.cs:1302]: If i < keys.Length once, it won't ever change right? For better at-a-glance-understanding™, I'd suggest creating some `bool translatingOut = i < keys.Length` and using that bool. Or place a comment.

File: src/Tasks/GetSDKReferenceFiles.cs (hunk 2 above)
Comment (Forgind) [reply src/Tasks/GetSDKReferenceFiles.cs:1263]: This has the bonus of deleting a summary that isn't actually accurate. Turns out if a mapping isn't in the dictionary, it doesn't look for it; it just assumes it isn't there.

File: src/Tasks/GetSDKReferenceFiles.cs (hunk 4 above)
Comment (Forgind) [reply src/Tasks/GetSDKReferenceFiles.cs:1255]: I don't think this actually mattered, so I reverted it. See next comment for fiddling.

File: src/Tasks/GetSDKReferenceFiles.cs (hunk 5 above)
Comment (Forgind) [reply src/Tasks/GetSDKReferenceFiles.cs:1322]: I had to fiddle with a lot of these to make them accessible to the test class. Then I downgraded a couple because properties were more visible than their containing class, which isn't really good style, but the latter wasn't necessary. As long as something prevents it from being public/protected, I don't think it particularly matters.

File: src/Tasks/GetSDKReferenceFiles.cs (hunk 6 above)
Comment (Forgind) [reply src/Tasks/GetSDKReferenceFiles.cs]: True. I put the if statement first, which should technically be more performant, too, though not in a particularly meaningful way.

File: src/Tasks/GetSDKReferenceFiles.cs (hunk 5 above)
Comment (BenVillalobos) [reply src/Tasks/GetSDKReferenceFiles.cs:1316]: Ah, I hadn't reviewed the test so I didn't see _any_ need for it.

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 7]
```diff
@@ -1246,63 +1218,102 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
+                    objTranslator(translator, ref value);
+                    dictionary[key] = value;
+                }
+            }
+            else
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    translator.Translate(ref keys[i]);
+                    T value = dictionary[keys[i]];
```
Comment (KirillOsenkov) [inline src/Tasks/GetSDKReferenceFiles.cs:1307]: Can the dictionary change while the loop runs? If a value is removed, this may throw. Do you need to lock? Also perhaps its better to enumerate key value pairs to avoid looking every key up.

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 8]
```diff
@@ -1246,63 +1218,102 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
+                translator.Translate(ref hash);
+            }
+        }
+
+        private static void TranslateConcurrentDictionary<T>(ITranslator translator, ref ConcurrentDictionary<string, T> dictionary, ObjectTranslator<T> objTranslator)
+        {
+            int count = dictionary.Count;
+            translator.Translate(ref count);
+            string[] keys = dictionary.Keys.ToArray();
+            if (keys.Length == 0)
```
Comment (KirillOsenkov) [inline src/Tasks/GetSDKReferenceFiles.cs:1291]: is this effectively determining the translation direction (read/write)? If so, is there a more explicit way, perhaps asking the translator?

Comment (BenVillalobos) [reply src/Tasks/GetSDKReferenceFiles.cs:1291]: Looks like ITranslator has a `TranslationDirection Mode` we can take advantage of here.

File: src/Tasks/GetSDKReferenceFiles.cs (hunk 7 above)
Comment (Forgind) [reply src/Tasks/GetSDKReferenceFiles.cs:1307]: No—before this PR, there was only the property, and it had a getter but no setter, so it really shouldn't be a ConcurrentDictionary. It's been a ConcurrentDictionary since the initial code commit. Do you want me to change it to an ordinary dictionary?

Comment (KirillOsenkov) [reply src/Tasks/GetSDKReferenceFiles.cs]: up to you, if it feels safe

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 9]
```diff
@@ -1246,63 +1218,79 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
         /// Structure that contains the on disk representation of the SDK in memory.
         /// </summary>
         /// <remarks>This is a serialization format. Do not change member naming.</remarks>
```
Comment (rainersigwald) [inline src/Tasks/GetSDKReferenceFiles.cs:1220]:
```suggestion
```

Stale

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 10]
```diff
@@ -1023,7 +1012,7 @@ internal SDKInfo GetCacheFileInfoFromSDK(string sdkRootDirectory, string[] sdkMa
 
                 PopulateRedistDictionaryFromPaths(directoryToFileList, redistDirectories);
 
-                var cacheInfo = new SDKInfo(references, directoryToFileList, FileUtilities.GetPathsHash(directoriesToHash));
+                var cacheInfo = new SDKInfo(references.ToDictionary(kvp => kvp.Key, kvp => kvp.Value, StringComparer.OrdinalIgnoreCase), directoryToFileList.ToDictionary(kvp => kvp.Key, kvp => kvp.Value, StringComparer.OrdinalIgnoreCase), FileUtilities.GetPathsHash(directoriesToHash));
```
Comment (rainersigwald) [inline src/Tasks/GetSDKReferenceFiles.cs:1015]: Is there a perf impact of this synchronization? Could you instead store an `IDictionary` in the SDKInfo fields and on creation it's `ConcurrentDictionary`/on deserialization it's a plain dictionary?

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 11]
```diff
@@ -1246,63 +1218,79 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
-        [Serializable]
-        private class SDKInfo
+        internal class SDKInfo : ITranslatable
         {
-            // Current version for serialization. This should be changed when breaking changes
-            // are made to this class.
-            private const byte CurrentSerializationVersion = 1;
+            private Dictionary<string, SdkReferenceInfo> pathToReferenceMetadata;
+            private Dictionary<string, List<string>> directoryToFileList;
+            private int hash;
```
Comment (rainersigwald) [inline src/Tasks/GetSDKReferenceFiles.cs:1225]: Our/runtime's standard naming convention would make these:

```suggestion
            private Dictionary<string, SdkReferenceInfo> _pathToReferenceMetadata;
            private Dictionary<string, List<string>> _directoryToFileList;
            private int _hash;
```

**Review by rainersigwald — APPROVED** [review]
Looks ok as is but I'm pretty curious if we can avoid the ConcurrentDictionary/Dictionary stuff.

File: src/Tasks/GetSDKReferenceFiles.cs [hunk 12]
```diff
@@ -1219,90 +1205,87 @@ public SdkReferenceInfo(string fusionName, string imageRuntime, bool isWinMD, bo
...
+                    t.Translate(ref imageRuntime);
+                    t.Translate(ref isManagedWinmd);
+                    t.Translate(ref isWinmd);
+                    info.FusionName = fusionName;
+                    info.ImageRuntime = imageRuntime;
+                    info.IsManagedWinmd = isManagedWinmd;
+                    info.IsWinMD = isWinmd;
+                }, count => new Dictionary<string, SdkReferenceInfo>(count, StringComparer.OrdinalIgnoreCase));
+
+                translator.TranslateDictionary(ref _directoryToFileList, (ITranslator t, ref string s) => t.Translate(ref s), (ITranslator t, ref List<string> fileList) =>
```
Comment (BenVillalobos) [inline src/Tasks/GetSDKReferenceFiles.cs:1276]: Funny that the duplicate translation worked because the same method handles input/output. Good thing it's not a huge concern and fairly easy to catch.

I wonder if a certain type of test would catch this? (not blocking on this thought)

**Review by BenVillalobos — APPROVED** [review]
LGTM!

File: src/Tasks/GetSDKReferenceFiles.cs (hunk 12 above)
Comment (Forgind) [reply src/Tasks/GetSDKReferenceFiles.cs:1276]: Right, but it should work as long as it matches. It serialized the dictionary twice, then deserialized it twice. Setting it the second time overwrote the correct value with the correct value, so everything still worked.

That also means I think it would be very hard to write a test for it...the only two options I can think of are having some kind of counter on how many times we accessed/set each variable or having a check at the end for the total size of the serialized form. Neither is trivial, and it isn't a huge deal for something that gets hit as often as this.

Comment (BenVillalobos) [reply src/Tasks/GetSDKReferenceFiles.cs:1276]: > Setting it the second time overwrote the correct value with the correct value, so everything still worked.

I see, I was thinking maybe there'd be duplicate data somehow. Agreed that it's fine as is.
